jax_galsim/moffat.py

```diff
@@ -0,0 +1,476 @@
+        if "half_light_radius" in self.params:
+            return self.params["half_light_radius"]
+        else:
+            return self._r0 * jnp.sqrt(
+                jnp.power(1.0 - 0.5 * self._fluxFactor, 1.0 / (1.0 - self.beta)) - 1.0
+            )
+
+    @property
+    def fwhm(self):
+        """The FWHM of this `Moffat` profle."""
+        if "fwhm" in self.params:
+            return self.params["fwhm"]
+        else:
+            return self._r0 * (2.0 * jnp.sqrt(2.0 ** (1.0 / self.beta) - 1.0))
+
+    @property
+    def _norm(self):
+        """Normalisation f(x) (trunc=0)"""
+        return self.flux * (self.beta - 1) / (jnp.pi * self._fluxFactor * self._r0**2)
+
+    @property
+    def _knorm(self):
+        """Normalisation f(k) (trunc = 0, k=0)"""
+        return self.flux
+
+    @property
+    def _knorm_bis(self):
+        """Normalisation f(k) (trunc = 0; k=/= 0)"""
+        return (
+            self.flux
+            * 4.0
+            / (jnp.power(2.0, self.beta) * jnp.exp(jax.lax.lgamma(self.beta - 1.0)))
+        )
+
+    def __hash__(self):
+        return hash(
+            (
+                "galsim.Moffat",
+                self.beta,
+                self.scale_radius,
+                self.trunc,
+                self.flux,
+                self.gsparams,
+            )
+        )
+
+    def __repr__(self):
+        return (
+            "galsim.Moffat(beta=%r, scale_radius=%r, trunc=%r, flux=%r, gsparams=%r)"
+            % (self.beta, self.scale_radius, self.trunc, self.flux, self.gsparams)
+        )
+
+    def __str__(self):
+        s = "galsim.Moffat(beta=%s, scale_radius=%s" % (self.beta, self.scale_radius)
+        if self.trunc != 0.0:
+            s += ", trunc=%s" % self.trunc
+        if self.flux != 1.0:
+            s += ", flux=%s" % self.flux
+        s += ")"
+        return s
+
+    @property
+    def _maxk_untrunc(self):
+        """untruncated Moffat maxK"""
+        ## JEC 8/7/23: new code w/o while_loop. Notice that some test using GalSim code itself leads to NaN for beta>=5.5
+        ## (this should be investigated with GalSim Tests)
+
+        ## JEC 21/1/23: see issue #1208 in GalSim github as it seems there is an error
+        ## in the expression used.
+
+        ##The 2D Fourier Transform of f(r)=C (1+(r/rd)^2)^(-beta) leads
+        ## C rd^2 = Flux (beta-1)/pi (no truc)
+        ## and
+        ## f(k) = C rd^2 int_0^infty (1+x^2)^(-beta) J_0(krd x) x dx
+        ##      = 2 F (k rd /2)^(\beta-1) K[beta-1, k rd]/Gamma[beta-1]
+        ## with k->\infty asymptotic behavior
+        ## f(k)/F \approx sqrt(pi)/Gamma(beta-1) e^(-k') (k'/2)^(beta -3/2) with k' = k rd
+        ## So we solve f(maxk)/F = thr  (aka maxk_threshold  in  gsparams.py)
+        ## leading to the iterative search of
+        ## let alpha = -log(thr Gamma(beta-1)/sqrt(pi))
+        ## k = (\beta -3/2)\log(k/2) + alpha
+        ## starting with k = alpha
+        ##
```

Comment:
I have put most in the doc string. It is important to be aware of the underlaying computation and to point the issue in the GalSim github. This could be revisited later